### PR TITLE
MEP: Standard Linux Distribution ("distro")

### PR DESCRIPTION
**Pull request by afbjorklund**
See #9992

### CODE REVIEW

**Review by medyagh — COMMENTED** [review]
if you think this the right way to do, we could approve this and work on it for Q1 2021.
should we have a transition period where people who really love our Buildroot ISO keep using it ?

File: enhancements/proposed/20201230-standard-distro/standard-distro.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+* Making major changes to the new standard operating system
+* Support production deployments, still intended for learning
+
+## Design Details
+
+Use external system image and external packages, same as for KIC image.
+
+## Alternatives Considered
+
+Continue to support custom distro, instead of using a standard distro.
```
Comment (medyagh) [inline enhancements/proposed/20201230-standard-distro/standard-distro.md:37]: should add consider making our buildroot ISO a separate project on its own?

**Review by medyagh — CHANGES_REQUESTED** [review]

**Comment by afbjorklund** [comment]
>  should we have a transition period where people who really love our Buildroot ISO keep using it ?

Indeed, the plan was to have multiple ISO files (one being the default) - and use URL to select the other

File: enhancements/proposed/20201230-standard-distro/standard-distro.md (hunk 1 above)
Comment (afbjorklund) [reply enhancements/proposed/20201230-standard-distro/standard-distro.md:37]: Yes, good suggestion! (add that option, explicitly)

**Review by tstromberg — APPROVED** [review]

**Review by priyawadhwa — APPROVED** [review]

**Review by medyagh — CHANGES_REQUESTED** [review]
one i think  I liket his MEP to conver is the upgrade plan,

it seems to me the previous versions of minikube will not be able to upgrade to newer verison of minikube if we change the ISO's OS.
and they would need to do selete and start a new one.

if that this the case, we need to discuss it and possibliy roll out minikube 2 which will not support upgrading old versions.

this would mean minikube would have to add a mechanisim to detect old iso and and warn user to delete and start again.

**Comment by afbjorklund** [comment]
We have changed iso before, when we went from boot2docker to buildroot...
As long as the files on the /var disk image are compatible, it should be doable.

It's also not a blocker, even if it _should_ require users to delete their cluster ?
I think we need a prototype, before we can tell if an upgrade is posssible or not

**Comment by tstromberg** [comment]
Agreed. In the past it hasn't been necessary to ask users to delete their
old VM's. Given how few entry points minikube actually has into the
built-in userland, I don't foresee compatibility issues.

I can see upgrades potentially being something to work out, but outside of
the scope of this MEP.

minikube already handles Ubuntu (docker) and buildroot (ISO) with minimal
behavioral differences.

**Comment by afbjorklund** [comment]
Before we look into making it a "general" distribution, we need something for the upgrade path.
Not saying that this is the only way, but it was _one_ of the options (keep the distro, but external)

I was going to look into A/B booting, like this: https://source.android.com/devices/tech/ota/ab
Basically you have **two** installations, one active and one passive (where you do the upgrade)

Alternative to using a package manager.

----

This would be similar to minikube rebooting with a new ISO, but slightly more hard disk-centric.
Both require that you also have a third image, where you keep the data that you want to keep.

I also had a fourth* image, that was used for pulling the images from (like the preload tarball)
This allows upgrading the container storage, if needed. But mostly used for the _first_ installation.

And a fifth image for `/data`, but anyway.

**Comment by afbjorklund** [comment]
> Agreed. In the past it hasn't been necessary to ask users to delete their
> old VM's. Given how few entry points minikube actually has into the
> built-in userland, I don't foresee compatibility issues.

I also wanted to make some drastic _changes_ to the userland, like excluding the runtimes.

https://github.com/kubernetes/minikube/issues/9989

**Pull request merged**